src/main/java/uk/gov/ida/matchingserviceadapter/resources/MatchingServiceResource.java

```diff
@@ -41,7 +77,41 @@ public MatchingServiceResource(
     public Response receiveSoapRequest(Document attributeQueryDocument) {
         LOG.debug("AttributeQuery POSTED: {}", attributeQueryDocument);
 
-        MatchingServiceResponse matchingServiceResponse = matchingService.handle(new MatchingServiceRequestContext(attributeQueryDocument));
-        return responseGenerator.generateResponse(matchingServiceResponse);
+        AttributeQuery attributeQuery = unwrapAttributeQuery(attributeQueryDocument);
+        List<Assertion> assertions = assertionDecrypter.decryptAssertions(new EncryptedAssertionContainer(attributeQuery));
```

Comment:
This isn't really in the spirit that `ValidatedEncryptedAssertionContainer` was intended.

It was supposed to enforce that assertions were not decrypted unless they had come from something with a valid signature. It doesn't really work of course, because you can just do what you've done and create an instance of `ValidatedEncryptedAssertionContainer` before you've actually validated anything.

I think if you're going down the route of translating first and validating second it's worth considering whether `decryptAssertions` should just take a list of `EncryptedAssertions` instead of its current trick of trying to use Java's type system to make sure they come from a container that's had a signature validated on it.

Reply:
Swap around/interleave the AQR validation and then do assertion decryption. This will sort out the healthcheck signature validation too as we won't be able to check the assertions.isEmpty() until after validation. Probably just make decryptAssertions take the list of encrypted assertions.